python/pyspark/ml/param/__init__.py

```diff
@@ -152,13 +152,17 @@ def isDefined(self, param):
         return self.isSet(param) or self.hasDefault(param)
 
     @since("1.4.0")
-    def hasParam(self, paramName):
+    def hasParam(self, param):
         """
-        Tests whether this instance contains a param with a given
-        (string) name.
+        Tests whether this instance contains a param.
         """
-        param = self._resolveParam(paramName)
-        return param in self.params
+        if isinstance(param, Param):
+            return hasattr(self, param.name)
```

Comment:
If we support ```param``` of type ```Param```, we should not only check the ```hasattr(self, param.name)``` but also check ```self.uid == param.parent```. You can directly call ```_shouldOwn``` to do this work. It means if you provide a ```Param``` to check whether it belongs to the instance, you should both check ```uid``` and ```name```.
I vote to only support ```paramName``` that make consistent semantics with Scala.